Is equal

gracetian6 opened the pull request:
The Is Equal function takes in a vector of type T and returns a boolean indicator vector of whether each element is equal to `value`. 

This contains
- pseudo code (old version not with preconditions)
- mathematical proof (want comments on whether it makes sense)

Please look at [is_equal pdf](https://github.com/opendp/whitepapers/blob/is_equal/is_equal/is_equal.pdf) as you review the PR!

SalilVadhan commented on is_equal/is_equal.tex:
```diff
@@ -0,0 +1,155 @@
+% \section{Cast}
+
+\section{Is Equal} 
+Lines 40-56 of \url{https://github.com/opendp/opendp/blob/21-impute/rust/opendp/src/trans/cast.rs#L40-L56}
```
is this an old version of the Rust code?  It doesn't seem like it ensures that M is SymmetricDistance.  (I think many of the comments I made in my review of clamp.tex will also apply to this file, so I may not repeat them all.)

gracetian6 replied:
yup, good call. I edited it so it has new code now.

SalilVadhan commented on is_equal/is_equal.tex:
```diff
@@ -0,0 +1,155 @@
+
+
+\subsection{Conditions as Specified in Pseudo Code}
+\begin{itemize}
```
many of these will move back into the code, right?

SalilVadhan commented on is_equal/is_equal.tex:
```diff
@@ -0,0 +1,155 @@
+% \silvia{I think perhaps it is useful to include the function as an equation}
+
+\begin{lstlisting}[language=Python]
+def make_is_equal(value : T, metric):
```
rather than making metric an input, why not simply write input_metric=SymmetricDistance, output_metric=SymmetricDistance?

SalilVadhan commented on is_equal/is_equal.tex:
```diff
@@ -0,0 +1,155 @@
+
+\begin{theorem}
+For every setting of the input parameters to \texttt{make\_is\_equal}, the transformation returned by \texttt{make\_is\_equal} has the following properties
+\begin{enumerate}
```
need to add domain-metric compatibility

SalilVadhan commented on is_equal/is_equal.tex:
```diff
@@ -0,0 +1,155 @@
+
+\subsection{Proof}
+
+\begin{theorem}
```
See clamp.tex for suggestions on the style and fonts here.

SalilVadhan commented on is_equal/is_equal.tex:
```diff
@@ -0,0 +1,155 @@
+
+
+\item \textup{(Stability Guarantee).}
+We can assume $d_{in} \leq d_{out}$ because Relation($d_{in}$, $d_{out}$) = \texttt{True}. We consider the symmetric distance as input metric. Since vector inputs $v, w$ are $d_{in}$-close, then by definition the symmetric distance of the multisets is bounded by $d_{in}$: $\abs{MultiSet(v) \Delta MultiSet(w)} \leq d_{in}$ 
```
should put relation and MultiSet in the math mode but using fonts like mathrm, mathit, and/or texttt.  Otherwise LaTeX formats it as if it is a product of the variables R, e, l, a, t, i, o, n, and s.

SalilVadhan commented on is_equal/is_equal.tex:
```diff
@@ -0,0 +1,155 @@
+
+We want to show that $\abs{MultiSet(function(v)) \Delta MultiSet(function(w))} \leq d_{out}$. %\\$\abs{MultiSet(function(v))\Delta MultiSet(function(w))} = \norm{h_{is\_equal(v)} - h_{is\_equal(w)}}_1 \leq d_{out}$. 
+To do so, we consider the symmetric distance between the multiset transformations $function(v)$ and $function(w)$. We claim that for any pair of set elements $v^* \in Multiset(v)$ and $w^* \in Multiset(w)$ $$\abs{MultiSet(function(v)) \Delta MultiSet(function(w))} \leq \abs{MultiSet(v) \Delta MultiSet(w)}$$
+We consider 4 cases. 
```
this sentence doesn't parse.  It says "We claim that for any pair v^* and w^*, STATEMENT holds", but STATEMENT makes no reference to v^* and w^*.

SalilVadhan replied:
I think you are trying to use the Path Property for SymmetricDistance, but you need to do so formally (referring to the defs and lemmas in the proof_defns doc) if you want to do that.  Else you can follow the simplified proof outline I suggested in clamp.tex.

gracetian6 replied:
Got it, will look at path property to formalize this proof.

SalilVadhan replied:
You may find the proof strategy suggested in clamp.tex to be simpler than using the path property.

gracetian6 commented:
New edits of 7/26 include

- updated style that includes preconditions, post conditions, and \texttt style for words
- Domain Metric Compatability
- Changed stability relation proof to use histogram notation & triangle inequality

**I'm planning to update with a new round of edits** using the new proof outline Salil suggested for row by row transformations. Will need to push new changes for the make_row_by_row transformation and it's corresponding pseudocode and proofs. 

**PRIORITIZE: does the proof for appropriate output domain make sense?**
